src/test/java/uk/gov/pay/products/service/PaymentCreatorTest.java

```diff
@@ -86,10 +95,76 @@ public void shouldCreateASuccessfulPayment() throws Exception {
                 paymentId,
                 paymentAmount,
-                paymentNextUrl);
+                paymentNextUrl,
+                productReturnUrl);
+
+
+        when(productDao.findByExternalId(productExternalId)).thenReturn(Optional.of(productEntity));
+        when(RandomIdGenerator.randomUuid()).thenReturn(paymentExternalId);
+        when(publicApiRestClient.createPayment(argThat(is(productApiToken)), argThat(PaymentRequestMatcher.isSame(expectedPaymentRequest)))).thenReturn(paymentResponse);
+
+        Payment payment = paymentCreator.doCreate(productExternalId);
+
+        assertNotNull(payment);
+        assertNotNull(payment.getExternalId());
+        assertThat(payment.getGovukPaymentId(), is(paymentResponse.getPaymentId()));
+        assertThat(payment.getNextUrl(), is(paymentResponse.getLinks().getNextUrl().getHref()));
+        assertThat(payment.getAmount(), is(paymentResponse.getAmount()));
+        assertNotNull(payment.getLinks());
+        assertThat(payment.getLinks().size(), is(2));
+        assertThat(payment.getLinks().get(0).getMethod(), is("GET"));
+        assertThat(payment.getLinks().get(0).getHref(), is(PRODUCT_URL + "/v1/api/payments/" + payment.getExternalId()));
+        assertThat(payment.getLinks().get(1).getMethod(), is("GET"));
+        assertThat(payment.getLinks().get(1).getHref(), is(paymentResponse.getLinks().getNextUrl().getHref()));
+        assertThat(payment.getProductId(), is(productEntity.getId()));
+        assertThat(payment.getProductExternalId(), is(productEntity.getExternalId()));
+        assertThat(payment.getStatus(), is(SUCCESS));
+
+        PaymentEntity expectedPaymentEntity = createPaymentEntity(
+                paymentId,
+                paymentNextUrl,
+                productEntity,
+                SUCCESS,
+                paymentAmount);
+        verify(paymentDao).merge(argThat(PaymentEntityMatcher.isSame(expectedPaymentEntity)));
+    }
+
+    @Test
+    public void shouldCreateASuccessfulPayment_whenReturnUrlIsNotPresent() throws Exception {
+        int productId = 1;
+        String productExternalId = "product-external-id";
+        long productPrice = 100L;
+        String productName = "name";
+        String productReturnUrl = "https://return.url";
+        String productApiToken = "api-token";
+
+        String paymentId = "payment-id";
+        String paymentExternalId = "random-external-id";
+        Long paymentAmount = 50L;
+        String paymentNextUrl = "http://next.url";
+
+        ProductEntity productEntity = createProductEntity(
+                productId,
+                productPrice,
+                productExternalId,
+                productName,
+                "",
+                productApiToken);
+        PaymentRequest expectedPaymentRequest = createPaymentRequest(
+                productPrice,
+                productExternalId,
+                productName,
+                productReturnUrl + "/" + paymentExternalId);
```

Comment:
For this test it makes slight sense. But, just a pattern match would be sufficient than a exact match in these cases.

Reply:
It is because how `PaymentRequestChecker` is written. `isSame()` would fail if the two external ids weren't the same.